arangod/Replication2/ReplicatedLog/Algorithms.cpp

```diff
@@ -326,29 +326,162 @@ auto algorithms::updateReplicatedLog(LogActionContext& ctx, ServerID const& serv
+};
+
+auto operator<=(ParticipantStateTuple left, ParticipantStateTuple right) noexcept -> bool {
+  if (left.index < right.index) {
+    return true;
+  } else if (left.index==right.index) {
+    return left.id <= right.id;
+  } else {
+    return false;
+  }
+}
 
-auto algorithms::calculateCommitIndex(std::vector<IndexParticipantPair>& indexes,
-                          std::size_t quorumSize, LogIndex spearhead)
+
+/* Add flag for failed server to IndexParticipant struct
+ *
+ *  add struct for parameters:
+ *   replicationFactor
+ *   - writeConcern (==quorumSize)
+ *   - softWriteConcern (better name?)
+ *   -
+ *  TRI_ASSERT(indexes.size() == replicationFactor)
+ *  TRI_ASSERT(writeConcern <= softWriteConcern <= replicationFactor)
+ *
+ *   actualWriteConcern = max(writeConcern, x)
+ *     where x = min(replicationFactor - number of failedServers,
+ *                   softWriteConcern);
+ *
+ *  for purposes of computing actualwriteConcern, excluded and failed
+ *  servers are treated the same.
+ *
+ * writeConcern is the *minimum* number of confirmed copies necessary to
+ * proceed with a commit for any given LogIndex
+ *
+ * softWriteConcern allows for a number of failed servers to be tolerated
+ * for the purposes of commit
+ *
+ * replicationFactor is the (expected) number of participants in the
+ * replicated log
+ *
+ * */
+
+algorithms::CalculateCommitIndexOptions::CalculateCommitIndexOptions(
+  std::size_t replicationFactor,
+  std::size_t writeConcern,
+  std::size_t softWriteConcern) :
+  _replicationFactor(replicationFactor), _writeConcern(writeConcern),
+  _softWriteConcern(softWriteConcern) {
+  TRI_ASSERT(_writeConcern <= _softWriteConcern)
+      << "writeConcern > softWriteConcern " << _writeConcern << " > "
+      << _softWriteConcern;
+  TRI_ASSERT(_softWriteConcern <= _replicationFactor)
+      << "softWriteConcern > opt.replicationFactor " << _softWriteConcern
+      << " > " << _replicationFactor;
+}
+
+auto algorithms::calculateCommitIndex(std::vector<ParticipantStateTuple>& indexes,
+                                      CalculateCommitIndexOptions const opt,
+                                      LogIndex currentCommitIndex, LogIndex spearhead)
     -> std::pair<LogIndex, CommitFailReason> {
-  auto nth = indexes.begin();
-  std::advance(nth, quorumSize - 1);
+  TRI_ASSERT(indexes.size() == opt._replicationFactor)
+      << "number of participants != replicationFactor (" << indexes.size()
+      << " < " << opt._replicationFactor << ")";
+
+
+  // number of failed participants
+  auto nrFailedServers = std::count_if(std::begin(indexes), std::end(indexes),
+                                       [](auto& p) { return p.isFailed(); });
+  auto actualWriteConcern =
+      std::max(opt._writeConcern, std::min(opt._replicationFactor - nrFailedServers,
+                                          opt._softWriteConcern));
+  TRI_ASSERT(actualWriteConcern > 0)
+      << "actualWriteConcern has to be at least 1";
```

Comment:
I think a write concern of 0 is valid. Stupid, but valid. You immediately ack every new entry and you can only recover from failure, when all servers are health.

Thinking about that, this would be a nice settings for imports. To shovel the data in as fast as possible.